core/telemetry/src/worker/node.rs

```diff
@@ -114,19 +118,35 @@ where TTrans: Clone + Unpin, TTrans::Dial: Unpin,
 	/// Polls the node for updates. Must be performed regularly.
 	pub fn poll(mut self: Pin<&mut Self>, cx: &mut Context) -> Poll<NodeEvent<TSinkErr>> {
 		let mut socket = mem::replace(&mut self.socket, NodeSocket::Poisoned);
+		let connection_timeout = Duration::from_millis(500);
 		self.socket = loop {
 			match socket {
-				NodeSocket::Connected(mut conn) =>
+				NodeSocket::Connected(mut conn) => {
+					self.connection_start = Some(Instant::now());
 					match NodeSocketConnected::poll(Pin::new(&mut conn), cx, &self.addr) {
 						Poll::Ready(Ok(v)) => match v {}
-						Poll::Pending => break NodeSocket::Connected(conn),
+						Poll::Pending => {
+							let timed_out = match self.connection_start {
```

Comment:
You should use https://docs.rs/futures-timer/0.3.0/futures_timer/struct.Delay.html

THis will wake up the task when we hit the timeout.

Reply:
Thanks